File: server/protocol/protocol.go
package protocol

import (
	"encoding/binary"
	"errors"
	"io"
)

// TCPメッセージタイプ
const (
	MsgHello          = 0x01
	MsgHelloAck       = 0x02
	MsgPTTStart       = 0x03
	MsgPTTStartAck    = 0x04
	MsgPTTStartDeny   = 0x05
	MsgPTTStop        = 0x06
	MsgPTTNotify      = 0x07
	MsgPTTNotifyStop  = 0x08
	MsgGroupChange    = 0x09
	MsgGroupChangeAck = 0x0A
	MsgHeartbeat      = 0x0B
	MsgDisconnect     = 0x0C
)

// UDPパケットタイプ
const (
	UDPTypeAudio     = 0x00
	UDPTypeKeepalive = 0x01
)

// テストトーン予約セッションID
const SessionIDTestTone = 0xFF

// CtrlMsg はTCP制御メッセージを表す
type CtrlMsg struct {
	Type    uint8
	Payload []byte
}

// UDPHeader はUDPパケットヘッダー（6バイト固定）
type UDPHeader struct {
	SessionID uint8
	Flags     uint8  // bit7=type, bit3-0=group_id
	Seq       uint16
	Timestamp uint16
}

func (h *UDPHeader) PacketType() uint8 { return (h.Flags >> 7) & 0x01 }
func (h *UDPHeader) GroupID() uint8    { return h.Flags & 0x0F }

// Encode は [payload_len][type][payload] 形式にシリアライズする
func Encode(msg *CtrlMsg) []byte {
	buf := make([]byte, 2+len(msg.Payload))
	buf[0] = uint8(len(msg.Payload))
	buf[1] = msg.Type
	copy(buf[2:], msg.Payload)
	return buf
}

// ReadMsg はTCPストリームから1メッセージを読み込む
func ReadMsg(r io.Reader) (*CtrlMsg, error) {
	hdr := make([]byte, 2)
	if _, err := io.ReadFull(r, hdr); err != nil {
		return nil, err
	}
	payLen := int(hdr[0])
	payload := make([]byte, payLen)
	if payLen > 0 {
		if _, err := io.ReadFull(r, payload); err != nil {
			return nil, err
		}
	}
	return &CtrlMsg{Type: hdr[1], Payload: payload}, nil
}

// ParseHello はHELLOペイロード [4B:device_id][1B:group_id] をパースする
func ParseHello(payload []byte) (deviceID uint32, groupID uint8, err error) {
	if len(payload) < 5 {
		return 0, 0, errors.New("HELLO payload too short")
	}
	deviceID = binary.BigEndian.Uint32(payload[0:4])
	groupID = payload[4]
	return
}

// MakeHelloAck はHELLO_ACKメッセージを生成する
func MakeHelloAck(sessionID uint8) *CtrlMsg {
	return &CtrlMsg{Type: MsgHelloAck, Payload: []byte{sessionID}}
}

// MakePTTNotify はPTT_NOTIFYメッセージを生成する
func MakePTTNotify(sessionID uint8) *CtrlMsg {
	return &CtrlMsg{Type: MsgPTTNotify, Payload: []byte{sessionID}}
}

// MakePTTNotifyStop はPTT_NOTIFY_STOPメッセージを生成する
func MakePTTNotifyStop() *CtrlMsg {
	return &CtrlMsg{Type: MsgPTTNotifyStop, Payload: nil}
}

File: server/protocol.py
"""
制御メッセージ / UDPヘッダー 定義とエンコード・デコード。
デバイス側 protocol.h と対応。
"""
import struct

# 制御メッセージタイプ
MSG_HELLO            = 0x01
MSG_HELLO_ACK        = 0x02
MSG_PTT_START        = 0x03
MSG_PTT_START_ACK    = 0x04
MSG_PTT_START_DENY   = 0x05
MSG_PTT_STOP         = 0x06
MSG_PTT_NOTIFY       = 0x07
MSG_PTT_NOTIFY_STOP  = 0x08
MSG_GROUP_CHANGE     = 0x09
MSG_GROUP_CHANGE_ACK = 0x0A
MSG_HEARTBEAT        = 0x0B
MSG_DISCONNECT       = 0x0C

# 音声UDPパケットタイプ (flags bit7)
UDP_TYPE_AUDIO     = 0x00
UDP_TYPE_KEEPALIVE = 0x01


def encode(msg_type: int, payload: bytes = b"") -> bytes:
    """[1B: payload_len][1B: type][payload]"""
    return bytes([len(payload), msg_type]) + payload


def decode(data: bytes) -> tuple[int | None, bytes]:
    """Returns (msg_type, payload). msg_type=None on error."""
    if len(data) < 2:
        return None, b""
    payload_len = data[0]
    msg_type    = data[1]
    payload     = data[2 : 2 + payload_len]
    return msg_type, payload


def parse_hello(payload: bytes) -> tuple[int, int]:
    """Returns (device_id, group_id). Raises ValueError on bad payload."""
    if len(payload) < 5:
        raise ValueError("HELLO payload too short")
    device_id = struct.unpack_from("!I", payload, 0)[0]
    group_id  = payload[4]
    return device_id, group_id


def make_hello_ack(session_id: int) -> bytes:
    return encode(MSG_HELLO_ACK, bytes([session_id]))


def make_ptt_notify(session_id: int) -> bytes:
    return encode(MSG_PTT_NOTIFY, bytes([session_id]))


def make_ptt_notify_stop() -> bytes:
    return encode(MSG_PTT_NOTIFY_STOP)


def udp_packet_type(flags: int) -> int:
    return (flags >> 7) & 0x01


def udp_group_id(flags: int) -> int:
    return flags & 0x0F
